src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,646 @@
+      if ($new_node === null) {
+        continue;
+      }
+
+      // we know we're rewriting the node, so now we know we need the namespace
+      $found_namespaces[$namespace] = true;
+
+      // replace it in the ast
+      $root = $root->replace($node, $new_node);
+
+      // potentially change adjacent expressions to check for null instead of false
+      if ($replace_config['replace_false_with_null'] ?? false) {
+        $root = $this->maybeChangeFalseToNull($root, $new_node);
+      }
+    }
+
+    if (C\count($found_namespaces) === 0) {
+      return $root;
+    }
+
+    // add "use namespace" declarations at the top if they aren't already present
+    $declarations =
+      vec($root->getDescendantsOfType(INamespaceUseDeclaration::class));
+    list($hsl_declarations, $suffixes) =
+      $this->findUseDeclarations($declarations);
+
+    $count_before = C\count(Vec\unique($suffixes));
+
+    // add new suffixes to the current list of suffixes
+    $suffixes = Vec\concat($suffixes, Vec\keys($found_namespaces))
+      |> Vec\unique($$);
+
+    // added any new suffixes?
+    if (C\count($suffixes) !== $count_before) {
+
+      // get all declarations in hte script
+      $root_declarations = $root->getChildren()['declarations'];
+
+      // remove any current use statements for HH\Lib\* namespaces, we'll group them together
+      $new_root_declarations = $root_declarations->removeWhere(
+        ($node, $parents) ==> C\contains($hsl_declarations, $node),
+      );
+      $children = vec($new_root_declarations->getChildren());
+      // first child is the <?hh sigil, keep that first
+      $first = C\firstx($children);
+
+      // all later declarations will go after our new use statement
+      $rest = Vec\drop($children, 1);
+
+      // build a possibly grouped namespace use declaration
+      $new_namespace_use_declaration = $this->buildUseDeclaration($suffixes);
+
+      // insert the <?hh sigil, then the new use statement, then everything else
+      $items = Vec\concat(vec[$first, $new_namespace_use_declaration], $rest);
+      $new_declarations = EditableList::fromItems($items);
+
+      // rewrite root
+      $root = $root->replace($root_declarations, $new_declarations);
+    }
+    return $root;
+  }
+
+  protected function resolveIntegerArgument(EditableNode $node): ?int {
+    if ($node instanceof LiteralExpression) {
+      $expr = $node->getExpression();
+      if ($expr instanceof DecimalLiteralToken) {
+        return (int)$expr->getText();
+      }
+
+      // a literal 0 shows as octal
+      if ($expr instanceof OctalLiteralToken) {
+        return (int)$expr->getText();
+      }
+
+      return null;
+    } elseif (
+      $node instanceof PrefixUnaryExpression &&
+      $node->getOperator() instanceof MinusToken
+    ) {
+      $val = $this->resolveIntegerArgument($node->getOperand());
+      return $val !== null ? -1 * $val : null;
```

Comment:
Please add brackets - this is fine, but `? :`'s weirdness in PHP makes me nervous :p